### ISSUE TITLE
Local vars not restored properly after call pausable method and throw exception

### ISSUE THREAD

**Issue by dennis-wang** [kilim/kilim#49]
public class Exceptions extends Task {
    private String fWithOutMailBox() throws Pausable {
        throw new RuntimeException();  // no pausable method
    }

```
private String fWithMailBox() throws Pausable {
    new Mailbox<String>().get(1000);  // has pausable method
    throw new RuntimeException();
}

public void execute() throws Pausable {
    String foo = "data";
    try {
        foo = fWithOutMailBox();
    } catch (Throwable t) {
        System.out.println("foo=" + foo);
    }
    String foo2 = "data";
    try {
        foo2 = fWithMailBox();
    } catch (Throwable t) {
        System.out.println("foo2=" + foo2);
    }
}

public static void main(String args[]) {
    new Exceptions().start();
}
```

}
================expect result will be
foo=data
foo2=data
================while real result is
foo=data
foo2=null

**Comment by nqzero** [comment]
i was able to reproduce and will take a look

**Comment by dennis-wang** [comment]
@nqzero   wait for your answer, we use the way blow first, so we can use default value if exception throws
String foo2 = null;
    try {
        foo2 = fWithMailBox();
    } catch (Throwable t) {
        System.out.println("foo2=" + foo2);
    }
if(foo2 == null){
   foo2 = "data";
}

**Comment by nqzero** [comment]
i have a fix that handles your case and passes the current (somewhat limited) tests

https://github.com/nqzero/kilim/tree/issues.49

give it a try and let me know if it works. thanks again for the bug report and excellent test case

i assume you're working on the master branch with java 7, so i backported the fix to that version. i'm prepping a major new release of kilim (probably called version 2.0) and i'll include this there along with a few other bug fixes and a bunch of new feature

**Comment by dennis-wang** [comment]
@nqzero  thank you very much , i'll have a try with your branches, and I have a question, does your branch checked out from the master and then fix?

**Comment by kilim** [comment]
This looks like a correct fix. I too had the same thing in mind ... make a pausable invocation a BasicBlock of its own. The reason is that the current design doesn't correctly handle the implicit jumps to an exception handler from all statements under a try block. More specifically, it doesn't handle variable liveness calculation correctly. In the example submitted, the "foo2" variable is deemed to have been killed, and replaced with the value returned from the call. This is not correct, because an exception inside the call results in the old value of foo2 still being preserved.  By modeling the pausable invocation as its own separate basic block, this problem is avoided.

**Comment by nqzero** [comment]
@dennis-wang  yes, it's against master

@sriram-srinivasan  it solves this specific issue, but it there are some similar problems. any instruction that can throw a caught exception is problematic. for checked exceptions this is only method calls that declare it, but for runtime exceptions it's any method call and many instructions, eg accessing a field of a possibly null object

without fancier analysis, every instruction in a try block needs to be treated as a branch

the simplest thing (which i'm going to try today) is to treat a BB and any handler paths as being in parallel, ie that a value needs to be def in every  one to count as def for the sake of liveness. the downside is this is overly conservative, esp for narrow catches

**Comment by nqzero** [comment]
@sriram-srinivasan 

i've just pushed 2 commits to https://github.com/nqzero/kilim/tree/fiber

the first shows the issue (kilim/Exrun)
the second is the conservative fix

those commits are temp - i'll do a forced push over them at some point

**Comment by nqzero** [comment]
updated my fiber branch. it treats try blocks conservatively, ie assumes that any instruction may fail, but also tracks when a var is first defined (ie born). if it hasn't been born, it can't be live. this prevents trying to use a variable before the info needed is available (which was causing the weaver and some tests to fail)

i still need to add a test for the case we're guarding against

https://github.com/nqzero/kilim/commit/04f18063bd5b64bc1e51717f6054cdfcf17f49f6

**Comment by nqzero** [comment]
added a test: https://github.com/nqzero/kilim/commit/6554cdc72a74070c86add4878908998244c9c463

for a large try block with a checked exception and writes to many variables, this approach is overkill, ie it ignores the writes in calculating liveness, causing possibly many more restores than needed. the splitting of blocks after pausable calls should no longer be strictly needed, but it helps mitigate some of these extraneous restores so i'm leaning towards leaving it in

my goal has been to favor correctness over performance. but for a topology in which it takes several passes for bornness and liveness to converge, eg a loop with conditional assignments and a catch, i'm not 100% confident that the Value calculations will be available ("UNDEFINED type" errors)

@kilim see anything wrong with this approach or an easy way to solve the problem in a more elegant / efficient manner ?

**Comment by nqzero** [comment]
i cleaned up the fiber branch (removed the example and squashed the commits): https://github.com/nqzero/kilim/commit/bc4fe1f4a285b189ac302b839815868e94688d1c

the old work is tagged: https://github.com/nqzero/kilim/tree/issues.49.fiber

**Comment by kilim** [comment]
> 
> @sriram-srinivasan it solves this specific issue, but it there are some similar problems. any instruction that can throw a caught exception is problematic. for checked exceptions this is only method calls that declare it, but for runtime exceptions it's any method call and many instructions, eg accessing a field of a possibly null object
> 

No, I don’t believe it is a problem. If there are no pausable calls related to a BB, then one doesn’t need to worry about the generated code either, since there is none! If there is a pausable method, it is (now) its own BB, which takes care of the def-use problem right away.

> without fancier analysis, every instruction in a try block needs to be treated as a branch

**Comment by nqzero** [comment]
from ExCatch.java (on my fiber branch)

        double  d = fd, d2 = d+1;
        if (doPause)
            Task.sleep(50);
        try {
            ((Object) null).toString();
            d = d2;
        }
        catch (Throwable ex) {}
        verify(d);


afaict, prior to my conservative liveness change, the weaver would see the assignment "d = d2" in the try block, decide d is not live in that block (the only successor to the sleep block) and assume it didn't need to restore d. the assignment never occurs, because of the NPE, and verify fails. when i wrote that comment, i was speculating, but i went and confirmed it to the best of my ability

deleting the assignment causes verify to succeed

**Comment by kilim** [comment]
> from ExCatch.java (on my fiber branch)
> 
>     double  d = fd, d2 = d+1;
>     if (doPause)
>         Task.sleep(50);
>     try {
>         ((Object) null).toString();
>         d = d2;
>     }
>     catch (Throwable ex) {}
>     verify(d);
> 
> afaict, prior to my conservative liveness change, the weaver would see the assignment "d = d2" in the try block, decide d is not live in that block (the only successor to the sleep block) and assume it didn't need to restore d.

Your observation is correct, that the analysis would consider d as kill’d.  But it wouldn’t matter at all, because there is no pausable call within the try block, so there is nothing to restore. There will be no kilim-related code in the thrown block. In this example, verify _should_ fail.

**Comment by nqzero** [comment]
when verify is called, shouldn't d==fd ?

**Comment by sriram-srinivasan** [comment]
> On Oct 26, 2016, at 10:19 PM, nqzero <notifications@github.com> wrote:
> 
> when verify is called, shouldn't d==fd ?

Oh duh! Yes, you are absolutely right.

**Comment by nqzero** [comment]
found a bug in the calculation of born-ness for method args of type long and double - they require 2 slots each. i've pushed a new version to my fiber branch

**Comment by nqzero** [comment]
this is fixed in [kilim 2.0](https://github.com/nqzero/kilim) by a series of commits, the last of which is https://github.com/nqzero/kilim/commit/e64e6af61ba03bd7d9d90398ed125f7d4fc63359

for the related commits, search the log messages for the words: born, live and 49 (ie this issue)
